[WIP]group機能の実装

zorori777 opened the pull request:
# WHAT
グループのモデル、コントローラの作成
グループ作成機能の実装
グループ編集機能の実装
グループ作成時のメンバー追加機能の実装
グループ編集時にメンバー追加機能の実装

# WHY
groupとしての情報を追加したり、編集したり保存するため。

zorori777 committed:
[wip]group

zorori777 committed:
group作成機能

# WHAT
・グループ作成機能
・validation　name追加
・flash message
# WHY
・グループを作るため
・nameがないと作成させないようにするため

takebnn26 commented on app/controllers/groups_controller.rb:
```diff
@@ -0,0 +1,32 @@
+
+
+private
+  def setting_group
```
これいるかな？
どちらもnew,  create アクションで@groupに再代入してるね

takebnn26 commented on app/views/groups/new.html.haml:
```diff
@@ -0,0 +1,22 @@
+.chat-group-form
```
これのビューを見たいのでスクリーンショットのっけて！

zorori777 replied:
https://gyazo.com/c65abecaac0bdf0097c724d7319b4469

zorori777 committed:
[WIP]  group 作成機能２

# WHAT
 ・cssの記述
・controller内のsetting_groupの記述削除
# WHY
・@new createでも@groupに再代入するため。
・viewのスクリーンショット

takebnn26 commented on app/controllers/groups_controller.rb:
```diff
@@ -20,10 +20,6 @@ def edit
 
 
 private
-  def setting_group
```
これ消したら当然他のところも消すよね！
言われてその部分だけを修正するだけじゃなくて、自分で考えて修正しましょう！

zorori777 committed:
group機能３

# WHAT
before_action :setting_groupの記述を削除
# WHY
settin_userの記述を削除したため

zorori777 committed:
group編集機能４

# WHAT
 ・routingの変更
・messageのindexにlink_to記述、グループの名前を記す
・routingの変更より、viewを変更
・message_controllerにindexアクション、group_idを取得
・groups_controllerにてredirect_toのパスを変更
# WHY
　groupを編集するため

takebnn26 commented on app/views/groups/new.html.haml:
```diff
@@ -0,0 +1,22 @@
+.chat-group-form
+  %h1 新規チャットグループ
+  = form_for (@group), class: 'new_chat_group' do |f|
```
ここも共通の部分テンプレート使えるよ！

takebnn26 commented on app/models/group.rb:
```diff
@@ -1,7 +1,11 @@
 class Group < ApplicationRecord
 
+
```
空行けしちゃって！

takebnn26 replied:
1行あけてればいいよ！

takebnn26 commented on app/controllers/messages_controller.rb:
```diff
@@ -1,6 +1,14 @@
 def index
 end
 
+
```
ここも1行あけてればいいよ！

takebnn26 commented on app/controllers/messages_controller.rb:
```diff
@@ -1,6 +1,14 @@
 class MessagesController < ApplicationController
 
+  before_action :setting_group
```
先にonly: [:index]つけたほうがいいかも

zorori777 committed:
group編集機能　５

# WHAT
 ・message_controllerにてindex飲みの記述
・formの使用
・空行削除

takebnn26 commented on app/controllers/messages_controller.rb:
```diff
@@ -1,11 +1,10 @@
 class MessagesController < ApplicationController
 
-  before_action :setting_group
+  before_action :setting_group,only: [:index]
```
onlyの前は1マスあけよう

zorori777 replied:
かしこまりました！！！！

zorori777 committed:
group user 追加

# WHAT
 ・controllerの記述変更
・collectionでuserの追加
・viewでの表示
# WHY
・userを追加するため

takebnn26 commented on app/controllers/messages_controller.rb:
```diff
@@ -1,6 +1,14 @@
+  before_action :setting_group,only: [:index]
+
 def index
+  @groups = Group.all
```
current_userに関連するグループだけもってこよう！

takebnn26 commented on app/controllers/groups_controller.rb:
```diff
@@ -0,0 +1,47 @@
+      redirect_to group_messages_path(@group), notice: '編集を成功しました'
+    else
+      flash.now[:alert] = "編集を失敗しました"
+        render :edit
```
これ失敗時にいけてる？

takebnn26 commented on app/views/groups/_chatside.html.haml:
```diff
@@ -0,0 +1,16 @@
+        %i.fa.fa-pencil-square-o
+  %ul.chat-side-group
+    %p チーム
+    - @groups.each do |group|
```
ここからrender使って部分テンプレート化しよう

takebnn26 replied:
eachの部分も含めて最短の形で部分テンプレート化しよう

takebnn26 commented on app/views/messages/index.html.haml:
```diff
@@ -1,46 +1,24 @@
+          %h2= @group.name
+        %p.chat-header--left__member
+          Members:
+          - @group.users.each do |user|
```
ここはもっとスマートに名前まで出せるよ
pluckメソッドとjoinメソッド

takebnn26 commented on config/routes.rb:
```diff
@@ -1,6 +1,9 @@
-  resources :messages,only: :index
+  root 'groups#index'
+  resources :groups,only: [:new,:create,:edit,:update] do
+    resources :messages,only: :index
```
カンマとの間は前半角開けよ

takebnn26 replied:
拡張できるように
scope module: 'groups'
で囲ってみよう調べてみて

takebnn26 commented on db/migrate/20161209113427_create_group_users.rb:
```diff
@@ -1,8 +1,8 @@
     create_table :group_users do |t|
-      t.references :user,  null:false
-      t.references :group, null:false
+      t.references :user
```
この二つ外部キー制約ついてないねつけよう

takebnn26 commented:
formの部分テンプレートってプッシュしてたっけ？

zorori777 committed:
groups users 追加

# WHAT
 ・userの追加をする
# WHY
・user を追加するため

takebnn26 commented on app/models/user.rb:
```diff
@@ -1,13 +1,11 @@
   has_many :messages
-  has_many :group_users
   has_many :groups, through: :group_users
+  has_many :group_users
```
このあと空行開けよう

takebnn26 commented on app/controllers/groups/messages_controller.rb:
```diff
@@ -0,0 +1,14 @@
+class Groups::MessagesController < GroupsController
+
+  before_action :setting_group,only: [:index]
```
onlyの前半角開けよう

takebnn26 commented on app/controllers/groups/messages_controller.rb:
```diff
@@ -0,0 +1,14 @@
+
+  before_action :setting_group,only: [:index]
+
+def index
```
インデント揃えよう

takebnn26 commented on app/controllers/groups/messages_controller.rb:
```diff
@@ -0,0 +1,14 @@
+def index
+end
+
+private
```
インデント

takebnn26 commented on app/views/groups/_chatside.html.haml:
```diff
@@ -0,0 +1,13 @@
+    .chat-side-user-right
+      = link_to edit_user_path(current_user), class: "chat-side-user-right" do
+        %i.fa.fa-cog
+      = link_to new_group_path(@group),class: "chat-side-user-right" do
```
カンマのあと半角開けよう

zorori777 committed:
user 追加

# WHAT
userを追加する
# WHY
userを追加する為

zorori777 commented:
さっきのindexのインデントまたミスしました、すいません(ﾉД`)ｼｸｼｸ

takebnn26 commented on app/views/groups/messages/index.html.haml:
```diff
@@ -0,0 +1,26 @@
+          %h2= @group.name
+        %p.chat-header--left__member
+          Members:
+          = User.pluck(:name).join(", ")
```
これだとユーザー全部や、グループに関連するユーザー持ってきておくれ

takebnn26 commented on app/controllers/groups_controller.rb:
```diff
@@ -0,0 +1,47 @@
+class GroupsController < ApplicationController
+
+before_action :setting_group, only: [:edit,:update]
```
インデント　半角開ける

takebnn26 commented on app/controllers/groups_controller.rb:
```diff
@@ -0,0 +1,47 @@
+  end
+
+
+private
```
インデント　半角開ける

takebnn26 commented on app/controllers/users_controller.rb:
```diff
@@ -16,7 +16,7 @@ def update
 
   private
     def update_params
```
ここインデント揃えて、privateの直後空行

zorori777 committed:
user 追加

# WHAT
userを追加
# WHY
userを追加するため

narukami894 commented on app/assets/stylesheets/modules/group.scss:
```diff
@@ -0,0 +1,107 @@
+  .chat-group-form__action-btn {
+    display: inline-block;
+    padding: 12px 20px;
+    color: #fff;
```
whiteと#fffが混在しているから、統一しよう
scssなので色の変数を定義してあげるのがベスト

snowhork replied:
これは？

narukami894 commented on app/controllers/application_controller.rb:
```diff
@@ -6,7 +6,7 @@ class ApplicationController < ActionController::Base
-   devise_parameter_sanitizer.permit(:sign_up,keys:[:name])
-   end
+    def configure_permitted_parameters
+ devise_parameter_sanitizer.permit(:sign_up,keys:[:name])
```
sign_upとkeysの間、keysとnameの間に半角スペースあけたげよう

narukami894 commented on app/controllers/groups/messages_controller.rb:
```diff
@@ -0,0 +1,15 @@
+  private
+
+    def setting_group
+     @groups = current_user.groups
```
インデント統一しよう

narukami894 commented on app/controllers/groups_controller.rb:
```diff
@@ -0,0 +1,47 @@
+  private
+
+  def create_params
+    params.require(:group).permit(:name, user_ids:[])
```
半角スペース

takebnn26 replied:
user_ids:の後に半角スペース

narukami894 commented on app/controllers/users_controller.rb:
```diff
@@ -5,18 +5,19 @@ def edit
-        render :edit
-      end
+    if @user.update_with_password(update_params)
+       redirect_to root_path,notice: '編集成功しました'
```
半角スペース！

narukami894 commented on app/views/groups/index.html.haml:
```diff
@@ -0,0 +1,2 @@
+.chat
+  = render partial: "chatside" ,locals:{groups: @groups}
```
半角スペース

narukami894 commented on app/views/groups/messages/index.html.haml:
```diff
@@ -0,0 +1,26 @@
+          %h2= @group.name
+        %p.chat-header--left__member
+          Members:
+          = @group.users.pluck(:name).join(",  ")
```
good

narukami894 commented on config/routes.rb:
```diff
@@ -1,6 +1,10 @@
-  resources :messages,only: :index
-  resources :users,only: [:edit,:update]
+  root 'groups#index'
+  resources :groups, only: [:new, :create, :edit, :update] do
```
%i記法使ってあげよう
http://qiita.com/mogulla3/items/46bb876391be07921743

yusukeyamane replied:
これはあああああああああああああああああああああああああああああああああ！？！？！？？！！？

snowhork replied:
これは

zorori777 committed:
user 追加

# What
 インデントの直し
# Why
user追加するため

takebnn26 commented on app/controllers/users_controller.rb:
```diff
@@ -16,11 +16,12 @@ def update
-      params.require(:user).permit(:name, :email,:password, :current_password, :password_comfirmation)
-    end
+  def update_params
+    params.require(:user).permit(:name, :email,:password, :current_password, :password_comfirmation)
```
:email,のあと半角スペースや
これやるだけでだいぶ見やすいコードになるから気をつけていこう！

takebnn26 commented on app/views/groups/index.html.haml:
```diff
@@ -1,2 +1,2 @@
 .chat
-  = render partial: "chatside" ,locals:{groups: @groups}
+  = render partial: "chatside" , locals:{groups: @groups}
```
locals:　の後も半角スペース

takebnn26 commented on app/views/layouts/application.html.haml:
```diff
@@ -1,4 +1,3 @@
-!!!
```
ん？これ無くしちゃあかん気がする

zorori777 committed:
user 追加

# what
user追加
・インデント直し
# why
userを追加するため

yusukeyamane commented on app/views/groups/messages/index.html.haml:
```diff
@@ -0,0 +1,26 @@
+          %p.chat-main-message__message !!!
+    .chat-footer-input
+      .chat-footer-input__left
+        %textarea{placeholder:"Type a message"}
```
中括弧の後はスペース

zorori777 committed:
user 追加

# WHAT
・レイアウトの調整
# WHY
・user追加するため

snowhork commented on app/views/groups/_form.html.haml:
```diff
@@ -0,0 +1,17 @@
+= form_for (@group)do |f|
```
部分テンプレートでインスタンス変数を用いるのはよろしくないので、
`render 'form', group: @group`
で渡してあげてください

snowhork commented on app/views/groups/_group.html.haml:
```diff
@@ -0,0 +1,5 @@
+- @groups.each do |group|
```
ここも同様にインスタンス変数を渡すのはよろしくないですね。
あと_group.html.hamlという場合は、group（単数）の部分テンプレートを書くべきなので、L1は消して、
`render @groups` で勝手にeach回してくれます

snowhork commented on app/views/groups/_chatside.html.haml:
```diff
@@ -0,0 +1,12 @@
+    .chat-side-user-right
+      = link_to edit_user_path(current_user), class: "chat-side-user-right" do
+        %i.fa.fa-cog
+      = link_to new_group_path(@group), class: "chat-side-user-right" do
```
ここも同様にインスタンス変数を渡すのはよろしくないですね。

takebnn26 replied:
link_to new_group_path(group)

snowhork commented on app/views/groups/index.html.haml:
```diff
@@ -0,0 +1,2 @@
+.chat
+  = render partial: "chatside" , locals:{ groups: @groups }
```
` = render 'chatside', groups: @groups`で良いですね

snowhork commented on app/views/groups/messages/index.html.haml:
```diff
@@ -0,0 +1,26 @@
+.chatmain
+  = render partial:'groups/chatside'
```
`= render 'groups/chatside'` ?

snowhork reviewed (commented):
pull req単位大きすぎるから次回からもうちょい細かく行こう！！！

zorori777 committed:
viewの変更

# what
 ・変更
# why
・userを追加するため！

zorori777 committed:
薬師寺さん指摘改善

# what
・routingの変更
・variable ファイルで色の変数定義
# why

takebnn26 commented on app/models/group.rb:
```diff
@@ -1,7 +1,7 @@
- has_many :group_users
- has_many :messages
+  has_many :users, through: :group_users
+  has_many :group_users
```
これthrough: :group_usersしてるならhas_many :group_usersを先に書いてあげよう

takebnn26 commented on app/controllers/application_controller.rb:
```diff
@@ -6,7 +6,7 @@ class ApplicationController < ActionController::Base
-   devise_parameter_sanitizer.permit(:sign_up,keys:[:name])
-   end
+    def configure_permitted_parameters
+ devise_parameter_sanitizer.permit(:sign_up, keys: [:name])
```
これインデントどうした！笑

takebnn26 replied:
devise_parameter_sanitizer.permit(:account_update, keys: [:name])
なくてもユーザー情報編集できたっけ？

zorori777 committed:
model controller modify

# WHAT
・表示の変更
# WHY
・user機能

zorori777 committed:
localchange

zorori777 committed:
revert

zorori777 committed:
Merge branch 'localchange' into group-product

zorori777 commented on app/assets/stylesheets/application.scss:
```diff
@@ -1,4 +1,5 @@
 @import "./reset";
+@import "variable";
```
ここお願いします！

zorori777 commented on app/controllers/application_controller.rb:
```diff
@@ -7,6 +7,7 @@ class ApplicationController < ActionController::Base
     def configure_permitted_parameters
- devise_parameter_sanitizer.permit(:sign_up, keys: [:name])
+      devise_parameter_sanitizer.permit(:sign_up, keys: [:name])
+      devise_parameter_sanitizer.permit(:account_update, keys: [:name])
```
ここおねがいします！

zorori777 commented on app/models/group.rb:
```diff
@@ -1,7 +1,9 @@
 class Group < ApplicationRecord
 
-  has_many :users, through: :group_users
   has_many :group_users
```
ここおねがいします

zorori777 commented on app/models/user.rb:
```diff
@@ -5,8 +5,9 @@ class User < ApplicationRecord
 
   has_many :messages
-  has_many :groups, through: :group_users
   has_many :group_users
```
ここおねがいします

zorori777 commented on config/routes.rb:
```diff
@@ -1,10 +1,10 @@
 Rails.application.routes.draw do
   devise_for :users
   root 'groups#index'
```
ここおねがいします！

takebnn26 commented on app/views/groups/_chatside.html.haml:
```diff
@@ -8,4 +8,5 @@
       = link_to new_group_path(@group), class: "chat-side-user-right" do
         %i.fa.fa-pencil-square-o
   %ul.chat-side-group
+    %p チーム
```
これいるんだっけ？

takebnn26 commented on config/routes.rb:
```diff
@@ -1,10 +1,10 @@
   devise_for :users
   root 'groups#index'
-  resources :groups, only: [:new, :create, :edit, :update] do
+  resources :groups, only: %i(new create edit update) do
```
exceptにしたほうがよさそう！

takebnn26 commented on db/migrate/20161228084535_add_name_to_users.rb:
```diff
@@ -1,5 +0,0 @@
-class AddNameToUsers < ActiveRecord::Migration[5.0]
```
インデックス貼ってる？

zorori777 committed:
user 追加

# what
 user 追加
# why
　user　追加

zorori777 commented:
index 貼ってあります！！
https://gyazo.com/0f44072bb1d19495a2bcde5465091a1c

takebnn26 commented on app/controllers/application_controller.rb:
```diff
@@ -6,7 +6,8 @@ class ApplicationController < ActionController::Base
-   def configure_permitted_parameters
-   devise_parameter_sanitizer.permit(:sign_up,keys:[:name])
-   end
+    def configure_permitted_parameters
```
ここのインデントはprivateと同じところでいいよ

takebnn26 commented on app/models/user.rb:
```diff
@@ -1,13 +1,13 @@
+  # :confirmable, :lockable, :timeoutable and :omniauthable
   devise :database_authenticatable, :registerable,
-         :recoverable, :rememberable, :trackable, :validatable
+          :recoverable, :rememberable, :trackable, :validatable
```
ここのインデントは上の:detabase_authenticatableに合わせよう

takebnn26 commented on app/views/groups/edit.html.haml:
```diff
@@ -0,0 +1,3 @@
+.chat-group-form
+  %h1 編集チャットグループ
+  = render 'form', group: @group
```
locals: { group: @group }

takebnn26 commented on app/views/groups/index.html.haml:
```diff
@@ -0,0 +1,2 @@
+.chat
+  = render 'chatside', groups: @groups
```
locals: {groups: @groups, group: @group}

takebnn26 commented on app/views/groups/new.html.haml:
```diff
@@ -0,0 +1,3 @@
+.chat-group-form
+  %h1 新規チャットグループ
+  = render 'form', group: @group
```
locals: {group: @group}

zorori777 committed:
user追加
